### PR TITLE
[libclc] Refactor build system to allow in-tree builds

### PR DESCRIPTION
**Pull request by frasercrmck**
The previous build system was adding custom "OpenCL" and "LLVM IR" languages in CMake to build the builtin libraries. This was making it harder to build in-tree because the tool binaries needed to be present at configure time.

This commit refactors the build system to use custom commands to build the bytecode files one by one, and link them all together into the final bytecode library. It also enables in-tree builds by aliasing the clang/llvm-link/etc. tool targets to internal targets, which are imported from the LLVM installation directory when building out of tree.

Diffing (with llvm-diff) all of the final bytecode libraries in an out-of-tree configuration against those built using the current tip system shows no changes. Note that there are textual changes to metadata IDs which confuse regular diff, and that llvm-diff 14 and below may show false-positives.

This commit also removes a file listed in one of the SOURCEs which didn't exist and which was preventing the use of ENABLE_RUNTIME_SUBNORMAL when configuring CMake.

### CODE REVIEW

**Review by rjodinchr — APPROVED** [review]
I have tested this patch to implement libclc build in-tree with clspv, and it worked perfectly!
Thanks for this awesome work

File: libclc/CMakeLists.txt [hunk 1]
```diff
@@ -27,31 +36,51 @@ set( LIBCLC_TARGETS_TO_BUILD "all"
 
 option( ENABLE_RUNTIME_SUBNORMAL "Enable runtime linking of subnormal support." OFF )
 
-find_package(LLVM REQUIRED HINTS "${LLVM_CMAKE_DIR}")
-include(AddLLVM)
+if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
```
Comment (rjodinchr) [inline libclc/CMakeLists.txt:39]: Maybe this should be an `OR` with `LIBCLC_STANDALONE_BUILD`, in order to have projects using libclc with `ExternalProject_Add` using that path? I'm not too sure

File: libclc/CMakeLists.txt [hunk 2]
```diff
@@ -358,3 +399,7 @@ foreach( t ${LIBCLC_TARGETS_TO_BUILD} )
     endif()
   endforeach( d )
 endforeach( t )
+
+if( NOT LIBCLC_STANDALONE_BUILD )
+  add_subdirectory( test )
```
Comment (rjodinchr) [inline libclc/CMakeLists.txt:404]: There is no `CMakeLists.txt` in `<llvm-project>/llvm/libclc/test`.

File: libclc/CMakeLists.txt (hunk 1 above)
Comment (frasercrmck) [reply libclc/CMakeLists.txt:39]: Yeah, I think you're right. I'm not 100%, but other LLVM sub-projects do it so I think we should too.

File: libclc/CMakeLists.txt (hunk 2 above)
Comment (frasercrmck) [reply libclc/CMakeLists.txt:404]: Ach yeah sorry, I have another path locally which adds some LIT tests - this should be in that one. I'll remove it for now, and we can maybe decide whether or not to add LIT testing to this PR or not.

**Comment by frasercrmck** [comment]
CC @mgorny -  I was unsure whether you are actively using and interested `libclc` or you were just looking after build bots or some other reason. I've successfully built the SPIR-V libraries out-of-tree but another pair of eyes can't hurt.

I was wondering - if we `libclc` building in-tree, would there be interest in building the SPIR-V targets with the LLVM SPIRV backend, rather than the external `llvm-spirv` tool?

**Review by rjodinchr — APPROVED** [review]

**Comment by mgorny** [comment]
Yeah, things seem to build and install fine for me too, with this patch applied.

File: libclc/CMakeLists.txt [hunk 3]
```diff
@@ -167,12 +181,14 @@ install( FILES ${CMAKE_CURRENT_BINARY_DIR}/libclc.pc DESTINATION "${CMAKE_INSTAL
...
-  add_library( subnormal_disable STATIC
-    generic/lib/subnormal_disable.ll )
-  install( TARGETS subnormal_use_default subnormal_disable ARCHIVE
-    DESTINATION "${CMAKE_INSTALL_DATADIR}/clc" )
+  foreach( file subnormal_use_default subnormal_disable )
+    link_bc(
+       TARGET ${file}
+       INPUTS ${PROJECT_SOURCE_DIR}/generic/lib/${file}.ll
+    )
+    install( FILES $<TARGET_PROPERTY:${file},TARGET_FILE> ARCHIVE
```
Comment (aaronpuchert) [inline libclc/CMakeLists.txt:189]: This causes an installation error for me with `-DENABLE_RUNTIME_SUBNORMAL:BOOL=ON`:
```
CMake Error at cmake_install.cmake:54 (file):
  file INSTALL cannot find
  "[...]/ARCHIVE": No such file or directory.
```
According to the [documentation](https://cmake.org/cmake/help/latest/command/install.html), `ARCHIVE` is a valid option for `install(TARGETS ...)`, but not for `install(FILES ...)`.

Comment (frasercrmck) [reply libclc/CMakeLists.txt:189]: See #109926

**Pull request merged**